samples/custom_dispatch/vulkan/shaders/pattern_module.mlir

```diff
@@ -0,0 +1,97 @@
+// RUN: iree-opt %s
+
+// The required configuration for the custom dispatch. This tells the compiler
+// the requisite target information needed to support the associated custom
+// shader.
+#spirv_target = #hal.executable.target<"vulkan", "vulkan-spirv-fb", {
+  spirv.target_env = #spirv.target_env<
+    #spirv.vce<v1.3, [Shader, GroupNonUniform], [SPV_KHR_storage_buffer_storage_class, SPV_KHR_variable_pointers]>,
+    #spirv.resource_limits<max_compute_workgroup_size = [128, 128, 64], subgroup_size = 64>
+  >
+}>
+
+#layout = #hal.pipeline.layout<push_constants = 1, sets = [
+  <0, bindings = [
+      <0, storage_buffer, ReadOnly>,
+      <1, storage_buffer, ReadOnly>,
+      <2, storage_buffer>
+  ]>
+]>
+#bindings = [
+  #hal.interface.binding<0, 0>,
+  #hal.interface.binding<0, 1>,
+  #hal.interface.binding<0, 2>
+]
+#objects = #hal.executable.objects<{
+  #spirv_target = [
+    #hal.executable.object<{
+      path = "samples/custom_dispatch/vulkan/shaders/simple_mul.spv"
+    }>
+  ]
+}>
+
+#attrdict = {
+  export = "main",
+  layout = #layout,
+  bindings = #bindings,
+  objects = #objects
+}
+
+module {
+  pdl.pattern : benefit(1) {
+    %lhs_type = pdl.type : tensor<?xf32>
+    %rhs_type = pdl.type : tensor<?xf32>
+    %out_type = pdl.type : tensor<?xf32>
+    %lhs = pdl.operand : %lhs_type
+    %rhs = pdl.operand : %rhs_type
+
+    // Match the target operation(s) for rewriting and the original arguments and result types.
+    %mul = pdl.operation "arith.mulf" (%lhs, %rhs : !pdl.value, !pdl.value) -> (%out_type : !pdl.type)
+
+    pdl.rewrite %mul {
+      // Constant attributes for rewriting. `attrdict` contains the layout/binding/object
+      // required for the custom dispatch.
+      %attrdict = pdl.attribute = #attrdict
+      %tied_operands = pdl.attribute = array<i64>
+      %c1_idx = pdl.attribute = 1 : index
+      %apply_map = pdl.attribute = affine_map<()[s0] -> (s0 ceildiv 64)>
+
+      %range = pdl.range %lhs : !pdl.value
+      %res_type_range = pdl.range %out_type : !pdl.type
+      %workload = pdl.apply_native_rewrite "get_tensor_sizes"(%range : !pdl.range<value>) : !pdl.range<value>
+      %dims = pdl.apply_native_rewrite "convert_index_to_i32"(%workload : !pdl.range<value>) : !pdl.range<value>
```

Comment:
note that i32 may not be sufficient for very large workloads - it's unfortunately just a reality here that we need to either make the decision to truncate (save limited bits - as we only have ~100bytes) on push constants but fail on big problems or extend to i64 as 2xi32 (wasting 32 bits in nearly all cases as we just pass 0 for the upper bits).

The best path forward is to eventually link SPIR-V modules with our own generated entry points so that we can optimize across them - this is what we do on the CPU side. The equivalent here would be to have the resource bindings and main() generated by the compiler and a .spv with just a function in it that takes everything as arguments instead of using the global state. Then we use the ukernel mechanism like the CPU side does to form dispatch regions around ukernels and everything else just works.

As you iterate on this path more that's something to keep in mind: instead of building out a lot more complex things it may be easier to implement ukernels in SPIR-V.

(this is still useful when wanting complete control, but will always be the hardest path due to the inability to change the ABI)

Reply:
For now I'll just leave the packing up to the user and can revisit once we see such larger workloads.